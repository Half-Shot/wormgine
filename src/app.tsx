import { useEffect, useMemo, useRef, useState } from 'preact/hooks'
import './app.css'
import { Game } from './game';

export function App() {
  const [game, setGame] = useState<Game>();
  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const game = new Game(window.innerWidth, window.innerHeight);
    game.loadResources().then(() => {
      setGame(game)
    });
  }, [false]);
 
  useEffect(() => {
    if (!ref.current || !game) {
      return;
    }
    if (ref.current.children.length > 0) {
      // Already bound
      return;
    }

    ref.current.appendChild(game.canvas);
    game.run();
  }, [ref, game]);

  return <div ref={ref} />;
}
