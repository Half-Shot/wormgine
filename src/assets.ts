import { Assets, Texture } from "pixi.js";
import {
  AssetData,
  AssetSounds,
  AssetTextures,
  manifest,
} from "./assets/manifest";
import { Sound } from "@pixi/sound";
import { BehaviorSubject, map } from 'rxjs';


let textures: Record<string, Texture>;
let sounds: Record<string, Sound>;
let data: Record<string, unknown>;

const internalAssetLoadPercentage = new BehaviorSubject(0);

export const assetLoadPercentage = internalAssetLoadPercentage.pipe();
export const assetsAreReady = internalAssetLoadPercentage.pipe(map<number, boolean>(v => v === 1));

export async function loadAssets() {
  await Assets.init({ manifest });

  const bundleCount = Object.keys(manifest.bundles).length;
  let bundleIndex = 0;
  for (const { name } of manifest.bundles) {
    const bundle = await Assets.loadBundle(name, (progress) => {
      const totalProgress = bundleIndex / bundleCount + progress / bundleCount;
      internalAssetLoadPercentage.next(totalProgress);
    });
    bundleIndex++;
    if (name === "textures") {
      textures = bundle;
    } else if (name === "sounds") {
      sounds = bundle;
    } else if (name === "data") {
      data = bundle;
    }
  }
}

export function getAssets() {
  if (!textures || !sounds || !data) {
    throw Error("Assets not preloaded");
  }
  return {
    textures: textures as unknown as AssetTextures,
    sounds: sounds as unknown as AssetSounds,
    data: data as unknown as AssetData,
  };
}

export type AssetPack = ReturnType<typeof getAssets>;
