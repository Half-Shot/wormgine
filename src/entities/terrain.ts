import Matter, { Composite, Body, Vector, Bodies, Collision, Detector, Query, Bounds, Vertices } from "matter-js";
import { UPDATE_PRIORITY, Container, DisplayObject, Graphics, ViewSystem, Rectangle } from "pixi.js";
import { IMatterEntity } from "./entity";

export class Terrain implements IMatterEntity {
    priority = UPDATE_PRIORITY.NORMAL;
    destroyed = false;

    private static SEGMENT_WIDTH = 5;
    private static TERRIAN_HEIGHT = 5;
    private static TERRAIN_VARIANCE = 15;

    private readonly gfx: Graphics = new Graphics();
    private readonly parts: Body[];

    get bodies() {
        return this.parts;
    }

    constructor(viewWidth: number, viewHeight: number, private readonly composite: Composite) {
        const width = viewWidth;
        const height = viewHeight - 200;

        // Draw the world
        const vertices: Vector[] = [];
        let i = 0;
        for (let x = 0; x < width + Terrain.SEGMENT_WIDTH; x += Terrain.SEGMENT_WIDTH) {
            const previousY = (vertices[i - 1]?.y ?? height);
            const variance = previousY + Math.round(Math.random() * Terrain.TERRAIN_VARIANCE) - (Terrain.TERRAIN_VARIANCE / 2);
            vertices.push(Vector.create(x, variance));
            i++;
        }
                
        const parts = [];
        for (let vI = 1; vI < vertices.length; vI++) {
            const last = vertices[vI - 1];
            const curr = vertices[vI];
            const newY = Math.min(last.y, curr.y);
            parts.push(Bodies.rectangle(last.x, newY, curr.x - last.x, Terrain.TERRIAN_HEIGHT, { isStatic: true }));
        }
        console.log(parts);
        this.parts = parts;

        // // Convert to simple bodies
        // this.terrainBody = Body.create({
        //     position: {
        //         y: height - 50,
        //         x: 6,
        //     },
        //     parts,
        //     isStatic: true,
        // });
    }

    async create(parent: Container<DisplayObject>, composite: Composite) {
        parent.addChild(this.gfx);
        Composite.add(composite, this.parts);
        //console.log(this.terrainBody.bounds);
    }

    update?(dt: number): void {
        this.gfx.clear();
        this.gfx.lineStyle(1, 0xFFBD01, 1);

        for (const rect of this.parts) {
            const gfxR = new Rectangle(rect.bounds.min.x, rect.bounds.min.y, rect.bounds.max.x - rect.bounds.min.x, rect.bounds.max.y - rect.bounds.min.y);
            this.gfx.drawShape(gfxR);
        }
    }

    onDamage(point: Vector, radius: number) {
        const explosionRadius = Bodies.circle(point.x, point.y, radius);
        this.gfx.lineStyle(1, 0xFFBD01, 1);
        this.gfx.drawCircle(point.x, point.y, radius);
        const collidedGround = Query.collides(explosionRadius, this.bodies);
        for (const collision of collidedGround) {
            const groundRect = collision.bodyA !== explosionRadius ? collision.bodyA : collision.bodyB;
            Composite.remove(this.composite, groundRect);
            this.bodies.splice(this.bodies.indexOf(groundRect), 1);
        }

        // TODO: Thoughts 2
        // - Remove anything intersecting explosionRadius
        // - Draw in new bodies around  the perimeter of the circle, using an algo of:
        // - Determine the impact points 
        // - Draw new parts along the radius of the circle between those two points.

        // Now, remove any colliding parts.
        // const explosionRadiusTimes2 = Bodies.circle(point.x, point.y, radius*2);
        // const allAffected = Query.region(this.bodies, explosionRadiusTimes2.bounds);
        // const toRemove = new Set<Body>();
        // for (const groundRect of allAffected) {
        //     // All the collision pieces
        //     for (const collision of Query.collides(groundRect, allAffected)) {
        //         if (collision.bodyA === collision.bodyB) {
        //             // Yes, this happens.
        //             continue;
        //         }
        //         const subA = Vector.magnitude(Vector.sub(point, collision.bodyA.position));
        //         const subB = Vector.magnitude(Vector.sub(point, collision.bodyB.position));

        //         console.log(subA, subB);

        //         if (subA > subB) {
        //             toRemove.add(collision.bodyB);
        //         } else {
        //             toRemove.add(collision.bodyA);
        //         }
        //     }
        // }

        // for (const body of toRemove) {
        //     Composite.remove(this.composite, body);
        //     this.bodies.splice(this.bodies.indexOf(body), 1);
        // }

        // // Finally, fill in gaps
        // const sortedRemaining = [...allAffected].filter(b => !toRemove.has(b)).sort(b => b.bounds.min.x);
        // for (let index = 0; index < sortedRemaining.length; index++) {
        //     const currentRect = sortedRemaining[index];
        //     const nextRect = sortedRemaining[index+1];
        //     if (!nextRect) {
        //         // At end;
        //         continue;
        //     }
        //     const gapX = Math.abs(nextRect.bounds.min.x - currentRect.bounds.max.x);
        //     const gapY = Math.abs(nextRect.bounds.min.y - currentRect.bounds.max.y);
        //     if (gapX > 10) {
        //         console.log('Found gap of', gapX, gapY);
        //     }
        // }
    }

    destroy(): void {
        throw new Error("Never destroyed.");
    }
}