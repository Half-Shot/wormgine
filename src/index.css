:root {
  font-family: Inter, system-ui, Avenir, Helvetica, Arial, sans-serif;
  line-height: 1.5;
  font-weight: 400;

  color-scheme: light dark;
  color: rgba(255, 255, 255, 0.87);
  background-color: #242424;

  font-synthesis: none;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-text-size-adjust: 100%;
}

@media (prefers-color-scheme: dark) {
  :root {
    --text: #bdbdbd;
    --bg: #2e2e2e;
    --highlight: #b96724;
    --links: #df8b1c;
    --subheading: #adadad;

    --team-red-bg: #cc3333;
    --team-red-fg: #db6f6f;
    --team-blue-bg: #2649d9;
    --team-blue-fg: #7085db;
    --team-purple-bg: #a226d9;
    --team-purple-fg: #bb70db;
    --team-yellow-bg: #d9c526;
    --team-yellow-fg: #dbcf70;
    --team-orange-bg: #d97a26;
    --team-orange-fg: #dba270;
    --team-green-bg: #30d926;
    --team-green-fg: #75db70;
    --team-unk-bg: #cc00cc;
    --team-unk-fg: #111111;
  }
}

a {
  font-weight: 500;
  color: #646cff;
  text-decoration: inherit;
}

a:hover {
  color: #535bf2;
}

body {
  margin: 0;
  display: flex;
  place-items: center;
}

#app {
  margin: 0;
  padding: 0;
  width: 100%;
}
