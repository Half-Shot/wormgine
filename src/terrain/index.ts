import { Vector2 } from "@dimforge/rapier2d-compat";
import { Rectangle } from "pixi.js";

export function imageDataToAlpha(
  boundaryX: number,
  boundaryY: number,
  imgData: ImageData,
): Map<number, Map<number, number>> {
  const data: Map<number, Map<number, number>> = new Map();
  const lengthOfOneRow = imgData.width * 4;
  for (let i = 0; i < imgData.data.length; i += 4) {
    const x = (i % lengthOfOneRow) / 4;
    const y = Math.ceil(i / lengthOfOneRow);
    const realX = x + boundaryX;
    const realY = y + boundaryY;
    const [, , , a] = imgData.data.slice(i, i + 4);
    if (!data.has(realY)) {
      data.set(realY, new Map());
    }
    data.get(realY)?.set(realX, a);
  }
  return data;
}

export function imageDataToTerrainBoundaries(
  alphas: ReturnType<typeof imageDataToAlpha>,
  imgData: ImageData,
): { boundaries: Vector2[]; boundingBox: Rectangle } {
  const boundingBox = new Rectangle(
    Number.MAX_SAFE_INTEGER,
    Number.MAX_SAFE_INTEGER,
    0,
    0,
  );
  const boundaries: Array<Vector2> = [];
  const xBoundaryTracker = new Array(imgData.width);
  const yBoundaryTracker = new Array(imgData.height);

  for (const [y, xValues] of alphas.entries()) {
    for (const [x, a] of xValues.entries()) {
      if (a > 5) {
        if (!xBoundaryTracker[x] || !yBoundaryTracker[y]) {
          // This is to stop us from drawing straight lines down
          // when we have a boundary, but I don't know why this works.
          if (x > 1 && y > 1) {
            boundaries.push(new Vector2(x, y));
          }
          xBoundaryTracker[x] = true;
          yBoundaryTracker[y] = true;

          boundingBox.x = Math.min(boundingBox.x, x);
          boundingBox.y = Math.min(boundingBox.y, y);
          boundingBox.width = Math.max(boundingBox.width, x);
          boundingBox.height = Math.max(boundingBox.height, y);
        }
      } else if (a === 0) {
        if (xBoundaryTracker[x] || yBoundaryTracker[y]) {
          boundaries.push(new Vector2(x, y));
          xBoundaryTracker[x] = false;
          yBoundaryTracker[y] = false;
        }
      }
    }
  }
  boundingBox.width -= boundingBox.x;
  boundingBox.height -= boundingBox.y;
  return {
    boundaries,
    boundingBox,
  };
}

export const QuadtreeCutoff = 4;

export function generateQuadsFromTerrain(
  boundaries: Vector2[],
  width: number,
  height: number,
  x: number,
  y: number,
): Rectangle[] {
  function inner(
    boundaries: Vector2[],
    width: number,
    height: number,
    x: number,
    y: number,
  ): Rectangle[] | Rectangle {
    // For performance, we just quad anything that's too small.
    if (width < QuadtreeCutoff || height < QuadtreeCutoff) {
      return new Rectangle(x, y, width, height);
    }

    // Are there any points within this quad?
    const interestedBoundaries = boundaries.filter(
      (v) => v.x >= x && v.x < x + width && v.y >= y && v.y < y + height,
    );

    // No? Turn it into a quad and stop recursing.
    if (interestedBoundaries.length === 0) {
      return new Rectangle(x, y, width, height);
    }

    // Split the quad into 4
    const newWidth = Math.round(width / 2);
    const newHeight = Math.round(height / 2);

    const rects: Rectangle[] = [];
    for (const opts of [
      [false, false],
      [true, false],
      [false, true],
      [true, true],
    ]) {
      const newX = x + (opts[0] ? newWidth : 0);
      const newY = y + (opts[1] ? newHeight : 0);
      const newRects = inner(
        interestedBoundaries,
        newWidth,
        newHeight,
        newX,
        newY,
      );
      // For each inner quad, delete any that contain none of the terrain.
      if (Array.isArray(newRects)) {
        rects.push(...newRects);
      } else if (boundaries.some((s) => newRects.contains(s.x, s.y))) {
        rects.push(newRects);
      }
    }
    return rects;
  }
  const result = inner(boundaries, width, height, x, y);
  return Array.isArray(result) ? result : [result];
}
