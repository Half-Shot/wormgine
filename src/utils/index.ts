import { Vector, Vector2 } from "@dimforge/rapier2d-compat";
export { Coordinate, MetersValue } from "./coodinate";



export function magnitude(v: Vector) {    
    return Math.sqrt(Math.pow(Math.abs(v.x), 2) + Math.pow(Math.abs(v.y), 2));
}

export function add(a: Vector, b: Vector) {
    return new Vector2(a.x + b.x, a.y + b.y);
}

export function sub(a: Vector, b: Vector) {
    return new Vector2(a.x - b.x, a.y - b.y);
}

export function mult(a: Vector, b: Vector|number) {
    const bVec = typeof b === "number" ? { x: b, y: b} : b;
    return new Vector2(a.x * bVec.x, a.y * bVec.y);
}

export function pointOnRadius(originX: number, originY: number, radians: number, radius: number): Vector2 {
    const x = Math.cos(radians)*radius;
    const y = Math.sin(radians)*radius;
    return new Vector2(originX + x, originY + y);
}