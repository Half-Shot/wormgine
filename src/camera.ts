import { Viewport } from "pixi-viewport";
import { Point } from "pixi.js";
import { MovedEvent } from "pixi-viewport/dist/types";
import Logger from "./log";
import { MetersValue } from "./utils";
import {
  combineLatest,
  debounceTime,
  map,
  Observable,
  Subscription,
} from "rxjs";

const logger = new Logger("ViewportCamera");

export enum CameraLockPriority {
  // Do not lock the camera to this object
  NoLock = 0,
  // Snap the camera to this object if the current player isn't local, but allow the user to move away.
  SuggestedLockNonLocal = 1,
  // Snap the camera to this object, but allow the user to move away.
  SuggestedLockLocal = 2,
  // Lock the camera to this object, but only suggest it to local players.
  LockIfNotLocalPlayer = 3,
  // Always lock the camera to this object.
  AlwaysLock = 4,
}

export interface LockableEntity {
  cameraLockPriority$: Observable<CameraLockPriority>,
  destroyed: boolean,
  sprite: {
    position: Point,
  }
}

export class ViewportCamera {
  private currentLock: {
    target: LockableEntity;
    priority: CameraLockPriority;
    isLocal: boolean;
  } | null = null;
  private userWantsControl = false;
  private lastMoveHash = Number.MIN_SAFE_INTEGER;
  private cameraSub?: Subscription;

  public get lockTarget(): LockableEntity | null {
    return this.currentLock?.target ?? null;
  }

  constructor(
    private readonly viewport: Viewport,
    private readonly clampY: MetersValue,
    physicalEntities: Observable<IteratorObject<LockableEntity>>,
    currentPlayableIsLocal: Observable<boolean>,
  ) {
    viewport.on("moved", (event: MovedEvent) => {
      if (
        event.type === "clamp-y" ||
        event.type === "clamp-x" ||
        event.type === "snap"
      ) {
        // Ignore, the director moved us.
        return;
      }
      if (this.userWantsControl === false) {
        this.userWantsControl = true;
        // Reset move hash, since the camera is under control.
        this.lastMoveHash = 0;
        logger.debug("Player took control");
      }
    });
    combineLatest([physicalEntities, currentPlayableIsLocal])
      .pipe(debounceTime(200))
      .subscribe(([entities, currentPlayableIsLocal]) => {
        this.updateEntitySet(entities, currentPlayableIsLocal);
      });
  }

  public snapToPosition(
    newTarget: Point,
    priority: CameraLockPriority,
    currentEntityIsLocal: boolean,
  ) {
    const targetXY: [number, number] = [newTarget.x, newTarget.y];

    // Short circuit skip move if it's cheaper not to.
    const newMoveHash = newTarget.x + newTarget.y;
    if (this.lastMoveHash === newMoveHash) {
      return;
    }
    this.lastMoveHash = newMoveHash;

    const clampYTo = this.clampY.pixels - this.viewport.screenHeight / 2;

    if (targetXY[1] > clampYTo) {
      logger.debug(
        "Clamped",
        targetXY[1],
        this.viewport.screenHeight,
        this.clampY.pixels,
      );
      targetXY[1] = clampYTo;
    } else {
      logger.debug("Set Y to", targetXY[1]);
    }

    switch (priority) {
      case CameraLockPriority.SuggestedLockNonLocal:
        if (this.userWantsControl) {
          return;
        }
        // Need a better way to determine this.
        if (!currentEntityIsLocal) {
          this.viewport.moveCenter(...targetXY);
        }
        break;
      case CameraLockPriority.SuggestedLockLocal:
        if (this.userWantsControl) {
          return;
        }
        this.viewport.moveCenter(...targetXY);
        break;
      case CameraLockPriority.LockIfNotLocalPlayer:
        if (!currentEntityIsLocal) {
          this.viewport.moveCenter(...targetXY);
        } else if (!this.userWantsControl) {
          this.viewport.moveCenter(...targetXY);
        }
        break;

      case CameraLockPriority.AlwaysLock:
        this.viewport.moveCenter(...targetXY);
        break;
    }
  }

  public updateEntitySet(
    entities: IteratorObject<LockableEntity>,
    isLocal: boolean,
  ) {
    this.cameraSub?.unsubscribe();
    logger.info("Recalculating entity set");

    const obs = entities
      // XXX: This type is actually wrong.
      .filter((entity) => entity.cameraLockPriority$)
      .map((entity) =>
        entity.cameraLockPriority$.pipe(
          map((priority) => ({ priority, entity })),
        ),
      );

    this.cameraSub = combineLatest([...obs]).subscribe((entities) => {
      logger.info(
        "New camera lock requested",
        entities.map((e) => ({
          priority: e.priority,
          ent: e.entity.toString(),
        })),
      );
      let nextTarget = this.currentLock;

      // Apply new value for currentLock before iterating.
      const currentEnt = entities.find(
        (e) => e.entity === this.currentLock?.target,
      );
      if (this.currentLock && !currentEnt) {
        // Ent is no longer found.
        nextTarget = null;
      } else if (nextTarget && currentEnt) {
        nextTarget.priority = currentEnt.priority;
      }

      for (const { entity, priority } of entities) {
        if (priority === CameraLockPriority.NoLock) {
          continue;
        }
        if (this.currentLock && this.currentLock?.priority >= priority) {
          logger.debug(
            "New lock is not higher priority than",
            this.currentLock.target.toString(),
            CameraLockPriority[priority],
          );
          // Skipping as higher priority exists.
          continue;
        }
        if (entity !== this.currentLock?.target) {
          // Reset user control.
          this.userWantsControl = false;
          logger.debug("New lock target", entity.toString());
          nextTarget = {
            target: entity,
            priority,
            isLocal,
          };
        } else {
          logger.debug("New lock target is same as last, ignoring");
          continue;
        }
      }
      this.currentLock = nextTarget;
    });
  }

  public update() {
    if (this.currentLock && !this.currentLock.target.destroyed) {
      this.snapToPosition(
        this.currentLock.target.sprite.position,
        this.currentLock.priority,
        this.currentLock.isLocal,
      );
    }
  }
}
