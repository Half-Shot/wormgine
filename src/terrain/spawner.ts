import { Point, Rectangle, Texture } from "pixi.js";
import { Vector } from "@dimforge/rapier2d-compat";
import { Coordinate, magnitude, shuffle, sub } from "../utils";
import {
  imageDataToTerrainBoundaries,
  generateQuadsFromTerrain,
  imageDataToAlpha,
} from ".";
import { BitmapTerrain } from "../entities/bitmapTerrain";
import { Scenario } from "../levels/scenarioParser";
import { TeamDefinition, TeamGroup } from "../logic/teams";
import { WormSpawnRecordedState } from "../entities/state/wormSpawn";
import { GameDebugOverlay } from "../overlays/debugOverlay";

interface SpawnerOpts {
  waterLevel: number;
  wormHeightBuffer: number;
  hazardPoints: Vector[];
}

export function determineLocationsToSpawn(
  quads: Rectangle[],
  alpha: ReturnType<typeof imageDataToAlpha>,
  { wormHeightBuffer, waterLevel, hazardPoints }: SpawnerOpts,
): Coordinate[] {
  const columns: Map<number, number[]> = new Map();

  for (const q of quads.filter(
    (quad) => quad.y < waterLevel, // Less than the water level.
  )) {
    // Add some clearance above the ground (-15)
    if (columns.has(q.x)) {
      columns.get(q.x)?.push(q.y - 15);
    } else {
      columns.set(q.x, [q.y - 15]);
    }
  }

  const allowedPoints: Vector[] = [];
  for (const [x, yvalues] of columns.entries()) {
    let yValue: number;
    while ((yValue = yvalues.pop()!)) {
      // Ignore if worm doesn't fit.
      if (yvalues.some((otherY) => yValue - otherY <= wormHeightBuffer)) {
        continue;
      }
      if (
        allowedPoints.some(
          (s) => Math.abs(magnitude(sub({ x, y: yValue }, s))) <= 80,
        )
      ) {
        // Ignore if position is too close to previous above.
        continue;
      }
      if (
        hazardPoints.some(
          (hazard) => Math.abs(magnitude(sub({ x, y: yValue }, hazard))) <= 30,
        )
      ) {
        // Ignore if position is too close to a hazard.
        continue;
      }
      if (alpha.get(yValue)?.get(x)) {
        // Point is inside terrain.
        continue;
      }
      // Check for ground clerance.
      const leftSide = new Rectangle(x - 120, yValue, 40, 30);
      if (!quads.some((q) => q.intersects(leftSide))) {
        continue;
      }
      const rightSide = new Rectangle(x + 120, yValue, 40, 30);
      if (!quads.some((q) => q.intersects(rightSide))) {
        continue;
      }

      allowedPoints.push({ x, y: yValue });
    }
  }

  return allowedPoints.map((v) => Coordinate.fromScreen(v.x, v.y));
}

export function getSpawnPoints(
  bitmap: Texture,
  objects: Scenario["objects"],
  teams: TeamDefinition[],
): WormSpawnRecordedState[] {
  const waterHeight = objects.find((o) => o.type === "wormgine.water")?.tra.y;
  if (!waterHeight) {
    throw Error("Map contains no water height marker");
  }
  const tmpCanvas = BitmapTerrain.drawToCanvas(bitmap);
  const context = tmpCanvas.getContext("2d")!;
  const imgData = context.getImageData(0, 0, tmpCanvas.width, tmpCanvas.height);

  const alphas = imageDataToAlpha(0, 0, imgData);
  const { boundaries } = imageDataToTerrainBoundaries(alphas, imgData);
  tmpCanvas.remove();
  // Turn it into a quadtree of rects
  const quadtreeRects = generateQuadsFromTerrain(
    boundaries,
    tmpCanvas.width,
    tmpCanvas.height,
    0,
    0,
  );
  const allSpawns = shuffle(
    determineLocationsToSpawn(quadtreeRects, alphas, {
      wormHeightBuffer: 40,
      waterLevel: waterHeight,
      hazardPoints: objects
        .filter((o) => o.type === "wormgine.mine")
        .map((o) => o.tra),
    }),
  );

  GameDebugOverlay.registerDebugData("spawns", "0x666600").push(
    ...allSpawns.map((c) => new Point(c.screenX, c.screenY)),
  );

  const spawns: { uuid: string; group: TeamGroup; x: number; y: number }[] = [];
  for (const worm of teams.flatMap((t) =>
    t.worms.map((w) => ({ ...w, group: t.group })),
  )) {
    if (!worm.uuid) {
      throw Error("Worm must have a uuid");
    }
    const nextSpawn = allSpawns.find(
      (nextSpawn) =>
        spawns.some(
          (existingSpawn) =>
            magnitude(sub(nextSpawn.toWorldVector(), existingSpawn)) < 5,
        ) === false,
    );
    if (!nextSpawn) {
      console.log(spawns, nextSpawn);
      throw Error("No place to spawn worm");
    }
    spawns.push({
      uuid: worm.uuid,
      group: worm.group,
      x: nextSpawn.worldX,
      y: nextSpawn.worldY,
    });
  }

  return spawns.map(
    (w) =>
      new WormSpawnRecordedState({
        type: "wormgine.worm_spawn",
        x: w.x,
        y: w.y,
        properties: {
          "wormgine.worm_uuid": w.uuid,
        },
      }),
  );
}
