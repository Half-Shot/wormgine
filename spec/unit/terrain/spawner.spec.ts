import { afterEach, beforeEach, describe, expect, test } from "@jest/globals";
import { determineLocationsToSpawn } from "../../../src/terrain/spawner";
import { imageDataToTerrainBoundaries, imageDataToAlpha, generateQuadsFromTerrain } from "../../../src/terrain/index";
import { Canvas, createCanvas, loadImage } from "@napi-rs/canvas";
import { renderTest } from "../../test-utils/graphics-test";
import { Coordinate } from "../../../src/utils";

const img = loadImage('./src/assets/levels/testingGround.png');

async function getImgData() {
  const terrainImg = await img;
  const canvas = createCanvas(terrainImg.width + 10, terrainImg.height + 10);
  const context = canvas.getContext("2d")
  context.drawImage(terrainImg, 10, 10);
  return { imgData: context.getImageData(0, 0, terrainImg.width, terrainImg.height), canvas };
}

describe('determineLocationsToSpawn', () => {
  let imgData: ImageData;
  let canvas: Canvas;
  let points: Coordinate[];
  beforeEach(async () => {
    const res = await getImgData();
    imgData = res.imgData;
    canvas = res.canvas;
    points = [];
  });

  afterEach(async () => {
    const ctx = canvas.getContext('2d');
    ctx.strokeStyle = 'red';
    ctx.fillStyle = 'red';
    ctx.lineWidth = 2;
    for (const point of points) {
      ctx.beginPath();
      ctx.moveTo(0, 600);
      ctx.lineTo(canvas.width, 600);
      ctx.stroke();
      ctx.beginPath();
      ctx.arc(point.screenX, point.screenY, 3, 0, 2 * Math.PI)
      ctx.fill();
    }
    await renderTest(canvas);
  });

  test('it does not choose a position near the water line', async () => {
    const waterLevel = 600;
    const alphas = imageDataToAlpha(0, 0, imgData);
    const quads = generateQuadsFromTerrain(imageDataToTerrainBoundaries(alphas, imgData).boundaries, imgData.width, imgData.height, 0, 0);
    const alpha = imageDataToAlpha(0, 0, imgData);
    points = determineLocationsToSpawn(quads, alpha, { waterLevel, wormHeightBuffer: 30, hazardPoints: [] });
    expect(points.every(p => p.worldY < waterLevel))
  })

  test('positioning is sane', async () => {
    const alphas = imageDataToAlpha(0, 0, imgData);
    const quads = generateQuadsFromTerrain(imageDataToTerrainBoundaries(alphas, imgData).boundaries, imgData.width, imgData.height, 0, 0);
    const alpha = imageDataToAlpha(0, 0, imgData);
    points = determineLocationsToSpawn(quads, alpha, { waterLevel: 600, wormHeightBuffer: 30, hazardPoints: [] });
    expect(points).toMatchSnapshot();
  });
});
