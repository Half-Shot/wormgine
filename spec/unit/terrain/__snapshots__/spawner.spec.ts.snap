// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`determineLocationsToSpawn positioning is sane 1`] = `
[
  Coordinate {
    "worldX": 31.15,
    "worldY": 24.75,
  },
  Coordinate {
    "worldX": 31.15,
    "worldY": 16.85,
  },
  Coordinate {
    "worldX": 35.25,
    "worldY": 24.75,
  },
  Coordinate {
    "worldX": 35.25,
    "worldY": 16.85,
  },
  Coordinate {
    "worldX": 35.25,
    "worldY": 7.1,
  },
  Coordinate {
    "worldX": 39.4,
    "worldY": 24.75,
  },
  Coordinate {
    "worldX": 39.4,
    "worldY": 16.85,
  },
  Coordinate {
    "worldX": 39.4,
    "worldY": 7.1,
  },
  Coordinate {
    "worldX": 43.5,
    "worldY": 24.75,
  },
  Coordinate {
    "worldX": 43.5,
    "worldY": 16.85,
  },
  Coordinate {
    "worldX": 43.5,
    "worldY": 7.1,
  },
  Coordinate {
    "worldX": 47.65,
    "worldY": 24.75,
  },
  Coordinate {
    "worldX": 47.65,
    "worldY": 16.85,
  },
  Coordinate {
    "worldX": 47.65,
    "worldY": 7.1,
  },
  Coordinate {
    "worldX": 22.5,
    "worldY": 24.75,
  },
  Coordinate {
    "worldX": 26.65,
    "worldY": 24.75,
  },
  Coordinate {
    "worldX": 26.65,
    "worldY": 16.85,
  },
  Coordinate {
    "worldX": 51.75,
    "worldY": 24.75,
  },
  Coordinate {
    "worldX": 51.75,
    "worldY": 16.85,
  },
  Coordinate {
    "worldX": 51.75,
    "worldY": 7.1,
  },
  Coordinate {
    "worldX": 55.9,
    "worldY": 24.75,
  },
  Coordinate {
    "worldX": 55.9,
    "worldY": 16.85,
  },
  Coordinate {
    "worldX": 55.9,
    "worldY": 7.1,
  },
  Coordinate {
    "worldX": 60,
    "worldY": 24.75,
  },
  Coordinate {
    "worldX": 60,
    "worldY": 16.85,
  },
  Coordinate {
    "worldX": 64.15,
    "worldY": 24.75,
  },
  Coordinate {
    "worldX": 64.15,
    "worldY": 16.85,
  },
  Coordinate {
    "worldX": 68.25,
    "worldY": 24.75,
  },
  Coordinate {
    "worldX": 72.4,
    "worldY": 24.75,
  },
]
`;
